### PR TITLE
LF: Change error exception for Arithmetic builtins

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR is a priliminary step toward making Arithmetic Builtin
throwing proper Daml Excetpions.

This is part of #8020.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala [hunk 1]
```diff
@@ -48,8 +48,10 @@ object SError {
   }
 
   /** Arithmetic error such as division by zero */
-  final case class DamlEArithmeticError(message: String) extends SErrorDamlException {
-    override def toString: String = message
+  final case class DamlEArithmeticError(builtinName: String, args: ImmArray[String])
+      extends SErrorDamlException {
+    override def toString: String =
+      s"Arithmetic Exception while evaluating $builtinName(${args.iterator.mkString(",")})."
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala:54]:
```suggestion
      s"ArithmeticException while evaluating $builtinName(${args.iterator.mkString(",")})."
```
or lowercase `exception` but this seems nicer to me

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -870,70 +826,67 @@ private[lf] object SBuiltin {
...
-        try {
-          x.divide(y, scale, roundingMode)
-        } catch {
-          case e: ArithmeticException =>
-            throw DamlEArithmeticError(e.getMessage)
-        }
-      rightOrArithmeticError("overflow/underflow", SBigNumeric.fromBigDecimal(result))
+      for {
+        scale <- SBigNumeric.checkScale(unchekedScale).toOption
+        roundingMode = unchekedRoundingMode.toInt
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:861]:
```suggestion
        roundingMode = Try(toIntExact(unchekedRoundingMode)).toOption
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -870,70 +826,67 @@ private[lf] object SBuiltin {
...
+        if unchekedRoundingMode == roundingMode
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:862]:
```suggestion
```

**Review by cocreature — APPROVED** [review]
lgtm thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala:54]: :+1:

Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala]: IMO this should be `ArithmeticError` (to match the Daml type) not `ArithmeticException`.

**Review by sofiafaro-da — APPROVED** [review]
Great work! 👍

**Review by nickchapman-da — APPROVED** [review]
Awesome!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala [hunk 4]
```diff
@@ -48,8 +48,10 @@ object SError {
   }
 
   /** Arithmetic error such as division by zero */
-  final case class DamlEArithmeticError(message: String) extends SErrorDamlException {
-    override def toString: String = message
+  final case class DamlEArithmeticError(builtinName: String, args: ImmArray[String])
+      extends SErrorDamlException {
+    override def toString: String =
+      s"ArithemticError while evaluating $builtinName(${args.iterator.mkString(",")})."
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SError.scala:54]:
```suggestion
      s"ArithmeticError while evaluating $builtinName(${args.iterator.mkString(",")})."
```

**Pull request merged**